lib/tune.cpp

```diff
@@ -801,7 +834,59 @@ namespace quda
           tunable.apply(stream); // do initial call in case we need to jit compile for these parameters or if policy tuning
 
           timer.start();
-          for (int i = 0; i < tunable.tuningIter(); i++) {
+          for (int i = 0; i < candiate_iterations; i++) {
+            tunable.apply(stream); // calls tuneLaunch() again, which simply returns the currently active param
+          }
+          timer.stop();
+          qudaDeviceSynchronize();
+          auto error = qudaGetLastError();
+
+          if (error != QUDA_SUCCESS) { // check we don't have a sticky error
+            qudaDeviceSynchronize();
+            if (qudaGetLastError() != QUDA_SUCCESS)
+              errorQuda("Failed to clear error state %s\n", qudaGetLastErrorString().c_str());
+          }
+
+          float elapsed_time = timer.last() / candiate_iterations;
+          if ((error == QUDA_SUCCESS) && (tunable.launchError() == QUDA_SUCCESS))
+            tc.pushCandidate(std::make_pair(elapsed_time, param));
+
+          if ((verbosity >= QUDA_DEBUG_VERBOSE)) {
+            if (error == QUDA_SUCCESS && tunable.launchError() == QUDA_SUCCESS) {
+              printfQuda("C   %s gives %s\n", tunable.paramString(param).c_str(),
+                         tunable.perfString(elapsed_time).c_str());
+            } else {
+              printfQuda("    %s gives %s\n", tunable.paramString(param).c_str(), qudaGetLastErrorString().c_str());
+            }
+          }
+          candidatetuning = tunable.advanceTuneParam(param);
+          tunable.launchError() = QUDA_SUCCESS;
+        }
+
+        if (tc.empty()) { errorQuda("Auto-tuning failed for %s with %s at vol=%s", key.name, key.aux, key.volume); }
+
+        constexpr float min_tune_time = 1e-5f;
+        const int tuneiterations = std::max(static_cast<int>(std::ceil(min_tune_time / tc.getBestTime())), 3);
```

Comment:
Like the candidate iterations, we should leave the `3` more exposed and not buried in this equation.

Reply:
Also, I think the default here should 10.  Given you have the timer cap, there's no reason not to use this.

Reply:
This is to make sure we *at least* do 3 iterations here. The timer cap is to make sure we do more iterations on fast running kernels. And `min_time_tune` should get a sensible default to make sure we don't pick up too much noise / timing inaccuracy.

Reply:
Ha, I read it backwards 😆